src/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs

```diff
@@ -71,99 +83,280 @@ public long BytesWritten
+        /// Commits the JSON text written so far which makes it visible to the output destination.
         /// </summary>
-        /// <param name="isFinalBlock">Let's the writer know whether more data will be written. This is used to validate
-        /// that the JSON written so far is structurally valid if no more data is to follow.</param>
-        /// <exception cref="InvalidOperationException">
-        /// Thrown when incomplete JSON has been written and <paramref name="isFinalBlock"/> is true.
-        /// (for example when an open object or array needs to be closed).
-        /// </exception>
-        public void Flush(bool isFinalBlock = true)
+        /// <remarks>
+        /// In the case of IBufferWriter, this advances the underlying <see cref="IBufferWriter{Byte}" /> based on what has been written so far.
+        /// In the case of Stream, this writes the data to the stream and flushes it.
+        /// </remarks>
+        public void Flush()
         {
-            if (isFinalBlock && !_writerOptions.SkipValidation && (CurrentDepth != 0 || _tokenType == JsonTokenType.None))
-                ThrowHelper.ThrowInvalidOperationException_DepthNonZeroOrEmptyJson(_currentDepth);
+            FlushHelper();
 
+            if (_stream != null)
+            {
+                FlushHelperStream();
+                _stream.Flush();
+            }
+
+            _memory = default;
+            BytesCommitted += BytesPending;
+            BytesPending = 0;
+        }
+
+        /// <summary>
+        /// Commits any left over JSON text that has not yet been flushed which makes it visible to the output destination.
+        /// </summary>
+        /// <remarks>
+        /// In the case of IBufferWriter, this advances the underlying <see cref="IBufferWriter{Byte}" /> based on what has been written so far.
+        /// In the case of Stream, this writes the data to the stream and flushes it.
+        /// </remarks>
+        /// <remarks>
+        /// The <see cref="Utf8JsonWriter"/> can still be re-used after disposing but all its state has been reset.
+        /// </remarks>
+        public void Dispose()
+        {
             Flush();
+            ResetHelper();
+        }
+
+#if BUILDING_INBOX_LIBRARY
+        /// <summary>
+        /// Asynchronously commits any left over JSON text that has not yet been flushed which makes it visible to the output destination.
+        /// </summary>
+        /// <remarks>
+        /// In the case of IBufferWriter, this advances the underlying <see cref="IBufferWriter{Byte}" /> based on what has been written so far.
+        /// In the case of Stream, this writes the data to the stream and flushes it.
+        /// </remarks>
+        /// <remarks>
+        /// The <see cref="Utf8JsonWriter"/> can still be re-used after disposing but all its state has been reset.
+        /// </remarks>
+        public async ValueTask DisposeAsync()
```

Comment:
This is not available down-level. So, if the user doesn't call FlushAsync, before dispose, dispose will flush synchronously for those who can't use the `await using` pattern.

Reply:
That's fine.